### PR TITLE
MDEV-34062 Implement innodb_log_file_mmap on 64-bit systems

### PR DESCRIPTION
**Pull request by dr-m**
- [x] *The Jira issue number for this PR is: MDEV-34062*
## Description
In MariaDB Server 10.11 (but not 10.6), copying the InnoDB write-ahead log (`innodb_logfile0`) is taking an extremely long time, copying at most 3 MiB/s from a RAM disk.

The problem appears to be connected to the rather small default value `innodb_log_buffer_size=2m` (`log_sys.buf_size`). If I specify `--innodb-log-buffer-size=1g`, then `xtrabackup_copy_logfile()` will process much more data per a system call, the backup would finish in reasonable time during my Sysbench workload, and the top of `perf report` would look more reasonable as well.

We can do better read the log file via memory-mapped I/O when available. A fallback for file based I/O will remain.

## Release Notes
On 64-bit systems, we will introduce a read-only Boolean parameter `innodb_log_file_mmap` that that will be `OFF` by default on most platforms. On Linux and FreeBSD the default is `innodb_log_file_mmap=ON`. When enabled, the InnoDB crash recovery as well as `mariadb-backup` will use memory-mapped I/O for reading the `ib_logfile0`, instead of regular file system I/O.
## How can this PR be tested?
The added parameter is lightly covered by the test `innodb.log_file_size_online`.

The code is covered by existing tests in `./mtr --suite=mariabackup`. With this fix, the backup during the following is expected to complete in a few seconds:
```sh
sysbench /usr/share/sysbench/oltp_update_index.lua --mysql-socket="$MYSQL_SOCK" --mysql-user="$MYSQL_USER" --mysql-db=test --tables=64 --table_size=100000 &
mariadb-backup -u"$MYSQL_USER" --socket="$MYSQL_SOCK" --backup --target-dir=/dev/shm/backup
```
In [MDEV-34062](https://jira.mariadb.org/browse/MDEV-34062) you can find results for testing with all 8 combinations of `innodb_log_file_mmap` (backup and server) and `innodb_log_file_buffering` (server) when the log resides on a SATA HDD, as well as results for testing the server with each setting.
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [x] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [ ] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

### CODE REVIEW

File: storage/innobase/buf/buf0flu.cc [hunk 1]
```diff
@@ -1885,6 +1915,35 @@ inline void log_t::write_checkpoint(lsn_t end_lsn) noexcept
...
+    const size_t ps{my_system_page_size};
+    ut_ad(buf_free == calc_lsn_offset(get_lsn()));
+    size_t offset{buf_free & ~(ps - 1)};
+    const size_t start_offset{calc_lsn_offset(checkpoint_lsn) & ~(ps - 1)};
+    if (offset == start_offset);
+    else if (offset < start_offset)
+    {
+      madvise(buf + offset, start_offset - offset, MADV_DONTNEED);
+      fallocate(log.m_file, FALLOC_FL_ZERO_RANGE, offset,
+                start_offset - offset);
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:1930]: Based on my tests on a HDD (see [MDEV-34062](https://jira.mariadb.org/browse/MDEV-34062)), it seems that the `fallocate`, which aims to prevent a read-on-write problem for a subsequent round of writing the circular `ib_logfile0`, is making things worse.

The `MADV_DONTNEED`, which is available on POSIX, seems to help in the `innodb_log_file_buffering=ON` (no `O_DIRECT`) case when the entire `ib_logfile0` apparently fits in the Linux file system or block cache. So, I would remove the references to `FALLOC_FL_ZERO_RANGE` and enable this code on all POSIX.

I agree that for many setups, enabling memory-mapped log writes may be a bad idea because page faults during writes would lead to reads of the log file. I think that we should offer this option nevertheless. Possibly, `madvise(MADV_SEQUENTIAL)` could help, but I can imagine that it could also request the entire `ib_logfile0` to be read into cache even when it is not needed.

Comment (dr-m) [reply storage/innobase/buf/buf0flu.cc]: 3d2ec55fd92078b4023ec39af50fcd30aaf8667c removes the references to `FALLOC_FL_ZERO_RANGE`.

In my tests, the impact of this read-before-write does not seem too bad. Besides, the combination of memory-mapped writes in `mariadbd` and memory-mapped reads in `mariadb-backup --backup` can give a significant performance improvement, depending on the workload.

Comment (dr-m) [reply storage/innobase/buf/buf0flu.cc]: @vaintroub says that the impact of read-before-write is bad on Microsoft Windows; 043745bd8df8954b8d449313dc4872d98c3ee4b6 includes some tweaks from his testing. So, we might want to limit the scope to reading the log, in `mariadb-backup` only.

File: extra/mariabackup/xtrabackup.cc [hunk 2]
```diff
@@ -3360,25 +3368,126 @@ static my_bool xtrabackup_copy_datafile(ds_ctxt *ds_data,
...
+
+      retry_count= 0;
+    }
+    else
+    {
+      if (metadata_to_lsn)
+      {
+        if (metadata_to_lsn <= recv_sys.lsn)
+          return false;
+      }
```
Comment (dr-m) [inline extra/mariabackup/xtrabackup.cc:3446]: Related to #3370, the `metadata_to_lsn` will have to be replaced with `metadata_last_lsn`.

File: storage/innobase/buf/buf0flu.cc (hunk 1 above)
Comment (dr-m) [reply storage/innobase/buf/buf0flu.cc]: ea8749820ddfc3f8f42c9477b3e95e959fec8a5d addresses the issue on Microsoft Windows by not enabling memory-mapped log access in `mariadb-backup` by default. On other 64-bit platforms than Windows, it will be enabled by default. On Linux, if one specified `MAP_POPULATE`, there should be synchronous read-ahead. But we of course do not do that; we prefer on-demand paging.

I think that `innodb_log_file_mmap=ON` can make sense on any platform when the `ib_logfile0` is relatively small (users prefer fast recovery times, do not care about write performance, or there are not too many writes).

**Comment by dr-m** [comment]
The top commit db576beedf2a197d16a43cff44959fc0a7ad6a00 is only for better testing this on our CI. It disables the PMEM interface, so that the new `mmap` logic will be exercised on Linux, and it enables the `mmap` logic by default on all platforms. (Except in `mariadb-backup` on Windows.)

File: storage/innobase/trx/trx0trx.cc [hunk 3]
```diff
@@ -1260,20 +1260,33 @@ static void trx_flush_log_if_needed(lsn_t lsn, trx_t *trx)
...
-  {
-    cb.m_callback = (void (*)(void *)) thd_decrement_pending_ops;
-    log_write_up_to(lsn, flush, &cb);
-  }
+#if defined HAVE_INNODB_MMAP && defined __linux__
+  if (!log_sys.is_mmap());
+  else if (!flush)
+    /* Starting with Linux 2.6.19, msync(MS_ASYNC) is a no-op
+    and we do not need to do anything special for non-durable writes. */
+    return;
```
Comment (dr-m) [inline storage/innobase/trx/trx0trx.cc:1268]: We will need such logic also on other platforms that implement memory-mapped log files.

An assertion is failing at the start of `log_write_up_to()` [on Windows](https://buildbot.mariadb.org/#/builders/234/builds/31540/steps/9/logs/stdio):
```
CURRENT_TEST: mariabackup.undo_upgrade
mysqltest: At line 11: query 'CREATE TABLE t1(a varchar(60)) ENGINE INNODB' failed: <Unknown> (2013): Lost connection to server during query
```
The call stack is as follows:
```
server.dll!log_write_up_to()[log0log.cc:1100]
server.dll!trx_flush_log_if_needed()[trx0trx.cc:1283]
server.dll!trx_t::commit_in_memory()[trx0trx.cc:1504]
server.dll!trx_t::commit_low()[trx0trx.cc:1610]
server.dll!trx_t::commit_persist()[trx0trx.cc:1623]
server.dll!trx_t::commit()[trx0trx.cc:1634]
server.dll!dict_stats_save()[dict0stats.cc:3461]
server.dll!dict_stats_update()[dict0stats.cc:4153]
server.dll!ha_innobase::create()[ha_innodb.cc:13224]
```
I think that we must avoid specifying a `completion_callback` when `!flush && log_sys.is_mmap()` .

Comment (dr-m) [reply storage/innobase/trx/trx0trx.cc]: cbd77f8685edbcaad3dc74994cc2963e261155a0 fixes this, and also ensures that the callback will actually be used when using regular system calls for log writes.

**Comment by dr-m** [comment]
As requested by @vaintroub, I am removing the logic to allow the regular `ib_logfile0` to be written via a memory mapping. For future reference, I created 0ae45445a065b0e635ab83f27f12bea1512da494 in case someone might want to enable that logic later.

File: storage/innobase/log/log0log.cc [hunk 4]
```diff
@@ -975,19 +1059,18 @@ void log_write_up_to(lsn_t lsn, bool durable,
...
   }
 #endif
+  ut_ad(!log_sys.is_mmap());
 
 repeat:
   if (durable)
   {
     if (flush_lock.acquire(lsn, callback) != group_commit_lock::ACQUIRED)
       return;
-    flush_lock.set_pending(log_sys.get_lsn());
```
Comment (vaintroub) [inline storage/innobase/log/log0log.cc:990]: seems like a random change. I tend to remove all "set_pendings" later, I doubt their usefulness, but maybe not in this unrelated patch.

File: storage/innobase/log/log0log.cc [hunk 5]
```diff
@@ -1000,9 +1083,7 @@ void log_write_up_to(lsn_t lsn, bool durable,
   }
 
   if (durable)
-  {
```
Comment (vaintroub) [inline storage/innobase/log/log0log.cc:1003]: According to style guide,  curly brackets around single statement are OK.
This is also unrelated to the goal of this patch.
Can we leave it as-is, please, for now?

**Review by vaintroub — CHANGES_REQUESTED** [review]

File: sql/sql_class.cc [hunk 6]
```diff
@@ -5223,10 +5223,11 @@ extern "C" MYSQL_THD thd_increment_pending_ops(MYSQL_THD thd)
   end of async operation (such as end of group commit
   write flush)
 
-  @param thd THD
+  @param t THD
 */
-extern "C" void thd_decrement_pending_ops(MYSQL_THD thd)
+extern "C" void thd_decrement_pending_ops(void *t)
```
[the remark this answers is not in the document]
Comment (dr-m) [reply sql/sql_class.cc:5228]: The function signature mismatch is undefined behaviour, and the type cast is violating `clang -Wcast-function-type-strict`. This change can be split out and fixed separately. It is related to #3490, but this code does not exist in its current target branch 10.5.

File: storage/innobase/trx/trx0trx.cc [hunk 7]
```diff
@@ -1237,7 +1237,7 @@ trx_finalize_for_fts(
 }
 
 extern "C" MYSQL_THD thd_increment_pending_ops(MYSQL_THD);
-extern "C" void  thd_decrement_pending_ops(MYSQL_THD);
+extern "C" void  thd_decrement_pending_ops(void*);
```
[the remark this answers is not in the document]
Comment (dr-m) [reply storage/innobase/trx/trx0trx.cc:1240]: This change is necessary for fixing runtime errors from `clang -fsanitize=undefined`, similar to ones fixed in #3322.

I paid attention to this, because we are changing the call in `trx_flush_log_if_needed()`. Even if #3490 is eventually completed and merged, the code in `trx_flush_log_if_needed()` will be different in this 10.11 branch and hence I think that we should actually correct this as part of this change.

File: sql/sql_class.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (vaintroub) [reply sql/sql_class.cc:5228]: I know what it does, but the good change would use lambda or wrap it into tiny static function, at the place where cast was previously.

File: storage/innobase/trx/trx0trx.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (vaintroub) [reply storage/innobase/trx/trx0trx.cc:1240]: No, it is not necessarily to fix it exactly like this.

File: storage/innobase/log/log0log.cc (hunk 4 above)
Comment (dr-m) [reply storage/innobase/log/log0log.cc:990]: I must have been thinking that `log_flush()` would set the value a little later, after we’d have done some `pwrite()` and `fdatasync()`. Now I realize that the purpose of this early assignment is to ensure that concurrent waiters on `flush_lock` would be given an early promise that something is going to be made durable up to a particular LSN.

I will retain this code and add a comment.

Comment (vaintroub) [reply storage/innobase/log/log0log.cc:990]: in theory this gave waiters chance to spinloop a bit, if the expected soon-to-be flush_lsn LSN is less than lsn parameter.
In practice though, no performance improvement was ever measured, and  this spin was never noticed by any profiler.

File: storage/innobase/log/log0log.cc [hunk 8]
```diff
@@ -177,22 +145,72 @@ void log_file_t::write(os_offset_t offset, span<const byte> buf) noexcept
...
+                      os_offset_t size)
 {
+  if (my_system_page_size > 4096)
+    return MAP_FAILED;
+# ifndef HAVE_PMEM
+  if (!log_sys.log_mmap)
+    return MAP_FAILED;
+# endif
+
+  const bool read_only=
```
[the remark this answers is not in the document]
Comment (dr-m) [reply storage/innobase/log/log0log.cc]: d71200c3c7303b0dff36af316fd203bd7fece1aa simplifies this.

File: storage/innobase/log/log0recv.cc [hunk 9]
```diff
@@ -3995,8 +4000,8 @@ void recv_sys_t::apply(bool last_batch)
     in ascending order of buf_page_t::oldest_modification. */
     log_sort_flush_list();
 
-#ifdef HAVE_PMEM
-  if (last_batch && log_sys.is_pmem())
+#if defined HAVE_INNODB_MMAP && !defined _WIN32
+  if (last_batch && log_sys.is_mmap())
     mprotect(log_sys.buf, len, PROT_READ | PROT_WRITE);
```
[the remark this answers is not in the document]
Comment (dr-m) [reply storage/innobase/log/log0recv.cc:4005]: d71200c3c7303b0dff36af316fd203bd7fece1aa fixes this. The mapping indeed needs to remain read-only for anything else than PMEM.

File: storage/innobase/trx/trx0trx.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (vaintroub) [reply storage/innobase/trx/trx0trx.cc:1240]: Can you please leave  it as-is, and I will fix the overly picky compiler myself. We can't convert everything to void*, to please picky tools.

File: sql/sql_class.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (vaintroub) [reply sql/sql_class.cc:5228]: Can you please leave it as-is, including the allegedly unsafe cast. I'll fix that cast without changing function signature, and without doing changes in 3 different places.

[the remark this answers is not in the document]
Comment (dr-m) [reply sql/sql_class.cc]: 4b49835e2642d2333128e58cdafb516784f35ed3 leaves this alone. You may want to read https://reviews.llvm.org/D134831 to understand the background.

File: storage/innobase/trx/trx0trx.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (dr-m) [reply storage/innobase/trx/trx0trx.cc]: If you add a wrapper function, it will likely incur some runtime overhead.

[the remark this answers is not in the document]
Comment (vaintroub) [reply storage/innobase/trx/trx0trx.cc]: I do not think so, not when it is trivial, static, inline, and only casts typed pointer parameter to void *. Compilers are smart.
Anyway it does not make any sense to convert everything to void*, because a tool incorrectly said so. This creates a WTF for the reader.

**Review by vaintroub — APPROVED** [review]
